### DISCUSSION TITLE
Is it possible to "map" over `tar_combine()`

### DISCUSSION THREAD

**Discussion by jdtrat** [ropensci/tarchetypes#82] [Help] [answered]
Hi! Happy New Year, and thank you so much for such a wonderful ecosystem of packages. My question may be related to #35, but I was wondering if there was a good way to use `tarchetypes::tar_combine()` for subsets of static branching targets created with `tarchetypes::tar_map()`.

In the example below, I am mapping over data (mtcars and iris), subsetting them, and would like to combine the results into one data frame at the end. Because they don't have the same columns, I can't call `dplyr::bind_rows()` on all of them. What I would like is to be able to combine the targets `first_mtcars` and `last_mtcars` as `combined_data_mtcars`; similarly, I want to combine `first_iris` and `last_iris` as `combined_data_iris`. Is this possible with static branching or would I need to explore dynamic branching here?

Thanks so much 😄

``` r
targets::tar_dir({
  targets::tar_script({
    mapped <- tarchetypes::tar_map(
      targets::tar_target(
        first,
        head(data)
      ),
      targets::tar_target(
        last,
        tail(data)
      ),
      values = tibble::tibble(
        data = rlang::syms(c("iris", "mtcars"))
        )
    )
    
    combined <- tarchetypes::tar_combine(
      combined_data,
      mapped,
      command = dplyr::bind_rows(!!!.x)
    )
    
    list(mapped, combined)
  })
  targets::tar_manifest()
})
#> # A tibble: 5 × 3
#>   name          command                                                  pattern
#>   <chr>         <chr>                                                    <chr>  
#> 1 last_mtcars   "tail(mtcars)"                                           <NA>   
#> 2 last_iris     "tail(iris)"                                             <NA>   
#> 3 first_iris    "head(iris)"                                             <NA>   
#> 4 first_mtcars  "head(mtcars)"                                           <NA>   
#> 5 combined_data "dplyr::bind_rows(first_iris = first_iris, first_mtcars… <NA>
```
<sup>Created on 2022-01-09 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by wlandau** [comment, marked as answer]
Thanks, @jdtrat. In your case, I think it would be convenient to put the `tar_combine()` target inside `tar_map()`. Sketch:

``` r
library(targets)
  
tar_script({
  library(tarchetypes)
  library(tidyselect)
  
  # https://armchairecology.blog/iris-dataset/
  # https://allisonhorst.github.io/palmerpenguins/
  library(palmerpenguins)
  
  subsets <- list(
    tar_target(
      first,
      head(data)
    ),
    tar_target(
      last,
      tail(data)
    )
  )
  
  combined <- tar_combine(
    combined,
    subsets,
    command = dplyr::bind_rows(!!!.x)
  )
  
  tar_map(
    list(subsets, combined),
    values = tibble::tibble(
      data = rlang::syms(c("mtcars", "penguins"))
    ),
    unlist = TRUE # Flattened list of targets with informative names.
  )
})

tar_visnetwork()
```

![](https://i.imgur.com/bJmSqQr.png)

<sup>Created on 2022-01-09 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

In case that doesn't fit the use case, here is an alternative. `tar_map(unlist = TRUE)` gives you a flattened list of mapped targets with informative names, and you can use the names to choose which targets to combine. `tidyselect::eval_select()` makes this convenient for large numbers of targets.

``` r
library(targets)
  
tar_script({
  library(tarchetypes)
  library(tidyselect)
  
  # https://armchairecology.blog/iris-dataset/
  # https://allisonhorst.github.io/palmerpenguins/
  library(palmerpenguins)
  
  mapped <- tar_map(
    tar_target(
      first,
      head(data)
    ),
    tar_target(
      last,
      tail(data)
    ),
    values = tibble::tibble(
      data = rlang::syms(c("mtcars", "penguins"))
    ),
    unlist = TRUE # Flattened list of targets with informative names.
  )
  
  combined_mtcars <- tar_combine(
    combined_mtcars,
    mapped[eval_select(ends_with("mtcars"), mapped)],
    command = dplyr::bind_rows(!!!.x)
  )
  
  combined_penguins <- tar_combine(
    combined_penguins,
    mapped[eval_select(ends_with("penguins"), mapped)],
    command = dplyr::bind_rows(!!!.x)
  )
  
  list(mapped, combined_mtcars, combined_penguins)
})

tar_visnetwork()
#> Warning message:
#> Targets and globals must have unique names. Ignoring global objects that conflict with target names: combined_mtcars, combined_penguins. Warnings like this one are important, but if you must suppress them, you can do so with Sys.setenv(TAR_WARN = "false").
```

![](https://i.imgur.com/2GUiL8N.png)

<sup>Created on 2022-01-09 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Reply by jdtrat** [reply]
This is exactly what I was looking for -- thank you so much @wlandau! The `eval_select()` trick is super helpful!
